src/librustc_lint/types.rs

```diff
@@ -427,42 +428,81 @@ enum FfiResult<'tcx> {
     },
 }
 
+fn is_zst<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, did: DefId, ty: Ty<'tcx>) -> bool {
+    tcx.layout_of(tcx.param_env(did).and(ty)).map(|layout| layout.is_zst()).unwrap_or(false)
+}
+
+fn ty_is_known_nonnull<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, ty: Ty<'tcx>) -> bool {
+    match ty.sty {
+        ty::FnPtr(_) => true,
+        ty::Ref(..) => true,
+        ty::Adt(field_def, substs) if field_def.repr.transparent() => {
+            if let AdtKind::Union = field_def.adt_kind() {
```

Comment:
Yes, this line is only needed because of RFC 2645, but my reasons for including it are a bit different.

RFC 2645 leaves it unspecified whether transparent `union`s participate in nonnull optimizations, and I don't intend to change that in this pull request. Continuing to emit the `improper_ctypes` warning in this situation makes it clear that the this is (still) unspecified.

Transparent `enum`s should be fine, though, since we can always safely optimize them just like a transparent `struct`. The feature will be gated in other parts of the code, and I don't think an additional gate check is necessary here for transparent `enum`s.

If people prefer, I can remove this `union` check from this pull request and move it into the implementation of RFC 2645.

Reply:
Ok, this makes sense. I think I'd prefer to only accept structs here for now and cover other transparent ADTs once they actually exist. You're right that it should probably Just Work for enums, but I'd like to have tests for that and prefer to not have dead, untestable code, even temporarily.

Reply:
Okay, I've modified this to only accept `struct`s for now. I'll revise this later to accept transparent `enum`s (either as a part of pull request #60463 (if this pull request is merged first), or as a later update to this pull request (if #60463 is merged first)).